Move /admin/api to /spi and PUT->PATCH

peterbroadhurst opened the pull request:
> Leaving this as a draft until the updates to FFTM are complete, because otherwise early adopters trying out the feature will not have a functioning combination.

- [x] Rename the `/admin/api/v1` API to `/spi/v1`
- [x] Include `namespace:` prefix on all request IDs to all plugins, so that FFTM can pass the NS back to find/update the op
  - [x] Blockchain
  - [x] Data Exchange
  - ~Shared Storage~ - no async ops
  - [x] Tokens
- [x] `PUT`s change to `PATCH`s, because there are race conditions between delete/update that mean PUT semantics wouldn't be safe without optimistic concurrency locking
- [x] Make sure all APIs FFTM needs to function are on this API, so it doesn't need to use two
  - [x] `GET` `/spi/v1/operations`
  - [x] `GET` `/spi/v1/operations/{nsopid}`
  - [x] `PATCH` `/spi/v1/operations/{nsopid}`
  - [x] `GET` `/spi/v1/namespaces`
  - [x] `GET` `/spi/v1/namespaces/{ns}`
- [x]  Continue to allow FFTM to use cross-namespace calls on the internal API
- [x] Add namespaces to the API that FFTM has

> Previously we discussed additional changes, to allow event listeners hosted in FFTM to be stateless - and avoid the
> need to have a REST API + LevelDB in FFTM to manage listeners/checkpoints. This worked out to be too complex
> to change right now. Specifically because:
> 1. BatchPin is not tracked as a listener
> 2. Token connectors use the APIs of EthConnect today to directly create event streams

Note thread in discord: https://discord.com/channels/905194001349627914/979733575496790086

awrichar commented on Makefile:
```diff
@@ -69,7 +69,7 @@ $(eval $(call makemock, internal/namespace,        Manager,            namespace
 $(eval $(call makemock, internal/assets,           Manager,            assetmocks))
 $(eval $(call makemock, internal/contracts,        Manager,            contractmocks))
-$(eval $(call makemock, internal/adminevents,      Manager,            admineventsmocks))
+$(eval $(call makemock, internal/spievents,      Manager,            spieventsmocks))
```
nit: indentation

awrichar commented on internal/database/sqlcommon/contractlisteners_sql.go:
```diff
@@ -42,6 +42,7 @@ var (
 		"options",
+		"state",
 		"created",
 	}
```
I expected to see an "updated" field here (it's in the database migration and in the query factory).

peterbroadhurst replied:
This is removed now

awrichar commented on internal/contracts/manager.go:
```diff
@@ -58,7 +58,10 @@ type Manager interface {
 	GetContractListenerByNameOrID(ctx context.Context, ns, nameOrID string) (*core.ContractListener, error)
 	GetContractListeners(ctx context.Context, ns string, filter database.AndFilter) ([]*core.ContractListener, *database.FilterResult, error)
+	GetContractListenerByIDGlobal(ctx context.Context, nameOrID string) (*core.ContractListener, error)
+	GetContractListenersGlobal(ctx context.Context, filter database.AndFilter) ([]*core.ContractListener, *database.FilterResult, error)
```
Global collections are going away very soon (everything needs to be retrieved within a namespace)

peterbroadhurst replied:
This is going (only the cross-namespace operations query will be retained at the end of this PR - which FFTM relies upon for the very short term)

awrichar commented on pkg/core/contract_listener.go:
```diff
@@ -37,6 +37,12 @@ type ContractListener struct {
 	Topic     string                   `ffstruct:"ContractListener" json:"topic,omitempty"`
 	Options   *ContractListenerOptions `ffstruct:"ContractListener" json:"options,omitempty"`
+	State     *fftypes.JSONAny         `ffstruct:"ContractListener" json:"state,omitempty"`
+}
```
Shouldn't this have an "Updated" timestamp?

peterbroadhurst replied:
Sorry, this one needs to come out before this merges. I had started adding the ability for FFTM to perform updates to contract listeners... before realizing that wasn't practical as a next step due to the direct creation of listeners in EthConnect today by tokens. So I will remove this field.

awrichar commented on internal/database/sqlcommon/operation_sql.go:
```diff
@@ -181,22 +181,29 @@ func (s *SQLCommon) UpdateOperation(ctx context.Context, ns string, id *fftypes.
+		update = update.Set("status", status)
+	}
+	if errorMsg != "" {
+		update = update.Set("error", errorMsg)
```
Looks like it is not possible to clear an error message once one is set. It's probably fine, but just pointing it out.

peterbroadhurst replied:
This is a good point. Now splitting out to an update DTO, we can have a string pointer to make this distinction.

peterbroadhurst commented:
ok - merge battled, and I've removed all the unnecessary bits around contract listeners. Also addressed the ability to set the error message to the empty string.

peterbroadhurst commented:
I've put this back in draft, as per thread https://discord.com/channels/905194001349627914/983341711742885948 we need to make the `id` passed to connectors when an operation is performed, include the `namespace` as well as the `UUID` of the operation.

So `0df3d864-2646-4e5d-8585-51eb154a8d23` becomes `default:0df3d864-2646-4e5d-8585-51eb154a8d23`

awrichar commented on internal/coremsgs/en_api_translations.go:
```diff
@@ -23,6 +23,7 @@ var (
 	APIParamsConfigRecordKeyUpdate          = ffm("api.params.configRecordKey.update", "The configuration key to update. This should use dot notation to reference a key documented in https://hyperledger.github.io/firefly/reference/config.html")
 	APIParamsConfigRecordKeyGet             = ffm("api.params.configRecordKey.get", "The configuration key to get. This should use dot notation to reference a key documented in https://hyperledger.github.io/firefly/reference/config.html")
 	APIParamsOperationIDGet                 = ffm("api.params.operationID.get", "The operation ID key to get")
+	APIParamsOperationNamespacedID          = ffm("api.params.spiOperationID", "The operation ID as passed to the connector when the operation was performed, including the 'ns!' prefix")
```
Prefix is now `ns:`, right?

awrichar commented on internal/operations/manager.go:
```diff
@@ -46,6 +46,7 @@ type Manager interface {
 	SubmitOperationUpdate(plugin core.Named, update *OperationUpdate)
 	TransferResult(dx dataexchange.Plugin, event dataexchange.DXEvent)
 	ResolveOperationByID(ctx context.Context, ns, id string, op *core.OperationUpdateDTO) error
```
Is this method used, or is it totally replaced by the new one below?

awrichar commented on internal/orchestrator/orchestrator.go:
```diff
@@ -899,5 +900,10 @@ func (or *orchestrator) SubmitNetworkAction(ctx context.Context, ns string, acti
 		return i18n.NewError(ctx, coremsgs.MsgUnrecognizedNetworkAction, action.Type)
 	}
-	return or.blockchain.SubmitNetworkAction(ctx, fftypes.NewUUID(), key, action.Type)
+	// TODO: This should be a new operation type
```
ack

awrichar reviewed (approved):
Looks good for a step forward that doesn't break core functionality 👍 Just noted some minor things inline.

peterbroadhurst commented on internal/apiserver/server.go:
```diff
@@ -300,6 +299,11 @@ func (as *apiServer) createMuxRouter(ctx context.Context, o orchestrator.Orchest
 	return r
 }
 
+func (as *apiServer) notFoundHandler(res http.ResponseWriter, req *http.Request) (status int, err error) {
```
Yep, we set it on the router here: 
https://github.com/hyperledger/firefly/blob/85f10558ef898cdef4e1b3eca7c7636e94e6c335/internal/apiserver/server.go#L298
